database/database_manager.rs

```diff
@@ -8,45 +8,48 @@ use std::{
     collections::HashMap,
     fs,
     path::{Path, PathBuf},
-    sync::{Arc, RwLock},
+    sync::{Arc, RwLock, RwLockReadGuard},
 };
 
 use itertools::Itertools;
+use resource::constants::{database::INTERNAL_DATABASE_PREFIX, server::SYSTEM_FILE_PREFIX};
 use storage::durability_client::WALClient;
 
 use crate::{database::DatabaseCreateError, Database, DatabaseDeleteError, DatabaseOpenError, DatabaseResetError};
 
-#[macro_export]
-macro_rules! internal_database_prefix {
-    () => {
-        "_"
-    };
-}
-
 #[derive(Debug)]
 pub struct DatabaseManager {
     data_directory: PathBuf,
     databases: RwLock<HashMap<String, Arc<Database<WALClient>>>>,
 }
 
 impl DatabaseManager {
-    pub fn new(data_directory: &Path) -> Result<Self, DatabaseOpenError> {
-        let databases = fs::read_dir(data_directory)
-            .map_err(|error| DatabaseOpenError::CouldNotReadDataDirectory {
-                path: data_directory.to_owned(),
-                source: Arc::new(error),
-            })?
-            .map(|entry| {
-                let entry = entry.map_err(|error| DatabaseOpenError::CouldNotReadDataDirectory {
+    pub fn new(data_directory: &Path) -> Result<Arc<Self>, DatabaseOpenError> {
+        let entries = fs::read_dir(data_directory).map_err(|error| DatabaseOpenError::CouldNotReadDataDirectory {
+            path: data_directory.to_owned(),
+            source: Arc::new(error),
+        })?;
+
+        let mut databases: HashMap<String, Arc<Database<WALClient>>> = HashMap::new();
+
+        for entry in entries {
+            let entry_path = entry
+                .map_err(|error| DatabaseOpenError::CouldNotReadDataDirectory {
                     path: data_directory.to_owned(),
                     source: Arc::new(error),
-                })?;
-                let database = Database::<WALClient>::open(&entry.path())?;
-                Ok((database.name().to_owned(), Arc::new(database)))
-            })
-            .try_collect()?;
+                })?
+                .path();
+
+            if entry_path.file_name().unwrap().to_string_lossy().starts_with(SYSTEM_FILE_PREFIX) {
```

Comment:
The stream-based `filter` made the code a total mess with `try_`s forced by `Database::open()?`, so I've rewritten it to a loop. Didn't introduce any initial `capacity` to the HashMap, not sure if we need to worry about it (I'd assume the base case to be 1 system and 1 user db, so it's a single memory allocation operation anyway).